apps/omg_child_chain/lib/omg_child_chain/block_queue/block_queue_submitter.ex

```diff
@@ -0,0 +1,227 @@
+    end
+  end
+
+  @doc ~S"""
+  Compute which blocks need to be submitted based on
+  the blocks already mined on the rootchain and the blocks
+  formed in the childchain. All blocks that have been formed
+  after the last block the rootchain has seen should be submitted.
+
+  ## Examples
+
+    In this example, the blocks are simple maps, while in the
+    actual code, they would be %BlockSubmission{} structs.
+
+      iex> BlockQueueSubmitter.get_blocks_to_submit(%{
+      ...>   blocks: %{1000 => %{num: 1000}, 2000 => %{num: 2000}},
+      ...>   formed_child_block_num: 2000,
+      ...>   gas_price_to_use: 1,
+      ...>   mined_child_block_num: 1000,
+      ...>   child_block_interval: 1000
+      ...> })
+      [%{gas_price: 1, num: 2000}]
+
+  """
+  @spec get_blocks_to_submit(blocks_to_submit_type()) :: [%BlockSubmission{}]
+  def get_blocks_to_submit(
+        %{
+          blocks: blocks,
+          child_block_interval: child_block_interval,
+          gas_price_to_use: gas_price_to_use,
+          mined_child_block_num: mined_child_block_num,
+          formed_child_block_num: formed_child_block_num
+        } = state
+      ) do
+    first_block_to_mine_num = first_block_to_mine_num(mined_child_block_num, child_block_interval)
+
+    _ =
+      BlockQueueLogger.log(:preparing_blocks, %{
+        first_block_to_mine_num: first_block_to_mine_num,
+        formed_child_block_num: formed_child_block_num
+      })
+
+    blocks
+    |> Enum.filter(pending_mining_filter_func(state))
+    |> Enum.map(fn {_blknum, block} -> block end)
+    |> Enum.sort_by(& &1.num)
+    |> Enum.map(&Map.put(&1, :gas_price, gas_price_to_use))
+  end
+
+  @doc ~S"""
+  Submits a block to the rootchain. The actual rootchain module to use
+  can be passed if needed.
+
+  ## Examples
+
+      iex> BlockQueueSubmitter.submit(
+      ...>   %BlockSubmission{hash: "success", nonce: 1, gas_price: 1},
+      ...>   OMG.ChildChain.FakeRootChain
+      ...> )
+      :ok
+
+  """
+  @spec submit(%BlockSubmission{}) :: :ok
+  def submit(%BlockSubmission{hash: hash, nonce: nonce, gas_price: gas_price} = submission, chain \\ RootChain) do
+    _ = BlockQueueLogger.log(:submitting_block, submission)
+    {:ok, newest_mined_blknum} = chain.get_mined_child_block()
+
+    hash
+    |> chain.submit_block(nonce, gas_price)
+    |> process_submit_result(submission, newest_mined_blknum)
+    |> return_result_and_log()
+  end
+
+  @doc ~S"""
+  Submits a block to the rootchain. The actual rootchain module to use
+  can be passed if needed.
+
+  ## Examples
+
+      iex> BlockQueueSubmitter.process_submit_result(
+      ...>   {:ok, "hash"},
+      ...>   %BlockSubmission{},
+      ...>   1
+      ...> )
+      :ok
+
+  """
+  @spec process_submit_result(submit_result_t(), BlockSubmission.t(), BlockSubmission.plasma_block_num()) ::
+          :ok | {:error, atom}
+  def process_submit_result(submit_result, submission, newest_mined_blknum) do
+    case submit_result do
```

Comment:
this to me is still a weird implementation... we return `:ok` and hide the implementation detail, that it doesn't really matter if the block gets submitted here or not. in all but `{:ok, _}` tuple, the block will get resubmitted with adjusted gas price...

Reply:
Not sure, what exactly do you find confusing here (this just interprets the block submission results from the eth node in the specific context we have), however, IIRC, the block is resubmitted with adjusted gas price not when submission fails but when it is doesn't get mined in the expected time-window. So it definitely can also happen on `{:ok, _}` in here. Gas price and re-submission seems to be out-of-scope of this sub-module.

EDIT: I just noticed the `@doc` looks incorrect (cc/ @T-Dnzt) and maybe this is what misled you?